### ISSUE TITLE
Querying non-existent single page by URI returns error instead of null

### ISSUE THREAD

**Issue by ravewebdev** [wp-graphql/wp-graphql#1639]
## Expected behavior

When querying a single page by `URI` and the provided `URI` does not exist, no error is thrown, and the returned data is null.

## Actual behavior

The following query works fine when either a) the `idType` is set to `URI` _and_ the `id` provided exists, or b) the `idType` is set to `DATABASE_ID` (whether or not the `id` exists).

```
query GET_PAGE_BY_ID($id: ID!, $idType: PageIdType = URI, $imageSize: MediaItemSizeEnum = LARGE) {
  page(id: $id, idType: $idType) {
    databaseId
    date
    slug
    uri
    title
    author {
      node {
        slug
        nickname
      }
    }
    featuredImage {
      node {
        altText
        sourceUrl(size: $imageSize)
      }
    }
  }
}
```

However, if the `idType` is set to `URI` and the `id` does _not_ exist, the following error is returned:

```
{
  "errors": [
    {
      "debugMessage": "Cannot return null for non-nullable field \"Page.databaseId\".",
      "message": "Internal server error",
      "extensions": {
        "category": "internal"
      },
      "locations": [
        {
          "line": 3,
          "column": 5
        }
      ],
      "path": [
        "page",
        "databaseId"
      ],
      ...
    }
  ],
  "data": {
    "page": null
  },
  "extensions": {
    "debug": []
  }
}
```

A similar query for a single post, passing `SLUG` for `idType` and a non-existent `id` does _not_ throw an error, instead returning null for data.

Tested both via headless frontend and directly in GraphiQL IDE (same error both places).

Using WP GraphQL v1.0.3 (also tested w/ v102 and v1.0.4 – which is not available via WP plugin repository, btw)

**Comment by westende** [comment]
I suspect this is caused by #1640. The query yields an empty result object while it should yield no results (i.e. page = null).

**Comment by ravewebdev** [comment]
This is still an issue in WP GraphQL v1.1.2 and is causing our app to return unreliable results when querying by post slug.

**Comment by ravewebdev** [comment]
So far the following fields throw similar errors, _only_ when querying by URI if the page doesn't exist:
- databaseId
- uri
- author
- featuredImage

The first two throw the "Cannot return null for non-nullable field" type error, whereas the latter two throw an error like the following (different for each field):
```
Argument 1 passed to WPGraphQL\\Connection\\Users::WPGraphQL\\Connection\\{closure}() must be an instance of WPGraphQL\\Model\\Post, instance of WPGraphQL\\Model\\PostType given
```

**Comment by jasonbahl** [comment]
Thanks for the updated info. Checking this out.

**Comment by jasonbahl** [comment]
@ravewebdev I'm not able to replicate this on v1.1.2. 

When I make a query using a URI that does not exist, I get a `null` response and no errors. 

![Screen Shot 2021-01-20 at 12 40 08 PM](https://user-images.githubusercontent.com/1260765/105225996-a297c780-5b1c-11eb-88f5-12df96a360b5.png)

I'm not fully convinced this doesn't mean that there's a bug here, but perhaps there's more information I'm missing? 

Can you double check for sure that the v1.1.2 has the code changes reflected in [this release](https://github.com/wp-graphql/wp-graphql/pull/1674/files).

Make sure that we're for sure running the same code? 

If we are for sure running the same code, then next steps would be to do some debugging [here]( https://github.com/wp-graphql/wp-graphql/blob/develop/src/Type/Object/RootQuery.php#L499-L549) and figure out what exactly is going wrong.

Does it seem to make any difference if the homepage / page on front, etc settings are set one way or another?

**Comment by ravewebdev** [comment]
@jasonbahl I think the issue is in Permalink Settings for some reason. We're using the custom structure `/blog/%postname%/` and are encountering this issue, but when I remove the `/blog` (`/%postname%/`), the query runs fine. Custom structures with `post_id` are fine (e.g., `/archives2/%post_id%`) but having that prefix with `postname` apparently breaks everything. Any idea why that might be happening?

**Comment by jasonbahl** [comment]
Ah! This is helpful to have more info about permalinks! I’ll play with some different variations and see what I can find out.

**Comment by rodrigo-arias** [comment]
Having the same issue with that permalink structure.

**Comment by kakaeriel** [comment]
I'm having the same issue when accessing `/page-slug/{any_string}`, if the URI exists in database. 
WPGraphQL response an error like this 

```
{
            "debugMessage": "Argument 1 passed to WPGraphQL\\Connection\\Users::WPGraphQL\\Connection\\{closure}() must be an instance of WPGraphQL\\Model\\Post, instance of WPGraphQL\\Model\\PostType given, called in /blog/wp-content/plugins/wp-graphql/src/Type/WPConnectionType.php on line 444",
            "message": "Internal server error",
            "extensions": {
                "category": "internal"
            },
            "locations": [
                {
                    "line": 38,
                    "column": 3
                }
            ],
            "path": [
                "page",
                "author"
            ]
        }
        ```

**Comment by justlevine** [comment]
Did a little sleuthing of my own. and it seems the issue is somewhere in [`NodeResolver.php::resolve_uri()`](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L75)

Using the following query:
```graphql
query {
  notReal: page(id: "/not-real/", idType: URI) {
    id
    databaseId
    uri
  }
  real: page(id: "/real/", idType: URI) {
    id
    databaseId
    uri
  }
}
```

## When the permalink structure **does not** have a fixed prefix:
(e.g. `/%category%/%postname%/`)

Value of `$this->wp->query_vars` used to choose the loader( [starting line 352](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L352) ):
- `notReal`:
```log
Array (
    [post_type] => page
    [archive] => 
    [nodeType] => Page
    [uri] => /not-real/
    [category_name] => not-real
)
```
- `real`:
```log
Array (
    [post_type] => page
    [archive] => 
    [nodeType] => Page
    [uri] => /real/
    [page] => 
    [pagename] => real
)
```

For whatever reason `notReal` gets matched as a category, but since it's not an instance of `WP_Term`, [`null` gets returned](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L444) **as we want**.

## When the permalink structure **has** a fixed prefix:
e.g. `/news/%category%/%postname%/`

Value of `$this->wp->query_vars` used to choose the loader( [starting line 352](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L352) ):
- `notReal`:
```log
Array (
    [post_type] => page
    [archive] => 
    [nodeType] => Page
    [uri] => /not-real/
)
```
- `real`:
```log
Array (
    [post_type] => page
    [archive] => 
    [nodeType] => Page
    [uri] => /real/
    [page] => 
    [pagename] => real
)
```

As you can see, here for some reason `notReal` doesn't get any other `$query_vars` added, so it:
1.  matches to [`isset( $this->wp->query_vars['post_type'] )`](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L450), 
2. and because [`$query_vars['page']` isnt set](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L465), 
3. it returns [a `WP_Post_Type` object](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L476), instead of **`null` like we want**.

**Comment by justlevine** [comment]
On further investigation, it seems the issue stems from the rewrite matching on [line 208](https://github.com/wp-graphql/wp-graphql/blob/8342be523a13d21353fa6d4970b6b3dbd8eaf6d6/src/Data/NodeResolver.php#L208).

Because `get_page_by_path()` returns `null`, 
1. no match is set for [`$this->wp->matched_rule`](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L227),
2. and nothing gets set to [`$perma_query_vars`](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L242).

## Tl;Dr
1. Because of the quirks of `WP_Rewrite`, when no permalink prefix is set a non-existent `urimatches the post category rewrite rules, and `NodeResolver::resolve_uri()` returns `null` since no category exists.
2. However, when a permalink prefix **is** set, the `uri` doesn't match any rewrite rules.
3. This causes the function to have no `$perma_query_vars` set, which ultimately means [a WP_Post_Type object is returned](https://github.com/wp-graphql/wp-graphql/blob/8342be523a13d21353fa6d4970b6b3dbd8eaf6d6/src/Data/NodeResolver.php#L476), instead of `null`.

Theoretically the fix is either:
- quick & dirty: add a check to ensure we're not trying to get a WP_Post_Type object before returning it (i'm not sure what the use case is for this fallback, so I cant suggest an implementation).
- ideally: refactor the rewrite matching logic to account for the graphql query type. querying a `page`, `post`, or `myCpt` shouldnt match a rewrite rule for another type. With this approach you probably also need to revisit the conditional logic for returning the correct loaders. (No idea if this is even possible, my understanding of WP rewrite rules is abysmal).

**Comment by justlevine** [comment]
In the interim, removing `$this->wp->query_vars['page']` from the `isset()` check [on Line 465](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/NodeResolver.php#L465) is working in my local env, but without knowing the use case for `uri` is set but `page` isnt => return `WP_Post_Type` object, I can't recommend this as a real solve.

**Comment by rodrigo-arias** [comment]
Thanks @justlevine for your research. Current development code worked for my use case with no side effects noted `if ( isset( $this->wp->query_vars['nodeType'] ) && 'Page' === $this->wp->query_vars['nodeType'] && isset( $this->wp->query_vars['uri'] ) )`

**Comment by justlevine** [comment]
This seems to have been fixed sometime before v1.13.5:

![image](https://user-images.githubusercontent.com/29322304/206176628-48f1bec9-42f3-4168-9430-8fd677172eb7.png)
Note the post uri is prefixed with `news`, `notReal` returns null, and `debug` is empty.

To prevent regressions, I added tests in https://github.com/wp-graphql/wp-graphql/pull/2659/commits/03d5e9518272e13d6ab7f02b12a45ff88435e24a
